Sketcher: Prevent bad constraint names

leuriato opened the pull request:
Whenever a constraint name contains a space (for example), it becomes impossible to reference it by name in an expression.

I added checks to ensure that constraint names only contain alphanumericals and underscores so that they can always be used in expressions.

Fixes: https://github.com/FreeCAD/FreeCAD/issues/19767

AjinkyaDahale commented on src/Mod/Sketcher/Gui/EditDatumDialog.cpp:
```diff
@@ -213,15 +213,21 @@ void EditDatumDialog::accepted()
+                    Gui::NotifyUserError(
+                        sketch,
+                        QT_TRANSLATE_NOOP("Notifications", "Value Error"),
+                        "Invalid constraint name (must only contain alphanumericals, underscores, "
```
This should (probably) also be enclosed with `QT_TRANSLATE_NOOP` or something related.

AjinkyaDahale replied:
Also, suggested to use "...alphanumericals and underscores", unless there are other acceptablr entities (don't just separate by comma).

AjinkyaDahale commented on src/Mod/Sketcher/Gui/TaskSketcherConstraints.cpp:
```diff
@@ -1224,12 +1224,17 @@ void TaskSketcherConstraints::onListWidgetConstraintsItemChanged(QListWidgetItem
-        std::string escapedstr = Base::Tools::escapedUnicodeFromUtf8(newName.c_str());
+        if (newName != Base::Tools::getIdentifier(newName)) {
+            Gui::NotifyUserError(
+                sketch, QT_TRANSLATE_NOOP("Notifications", "Value Error"),
```
Do you think there may be value in making a helper function that does this check, so as to avoid a mistake if this policy changes in the future?

AjinkyaDahale reviewed (changes requested):
Apart from these comments, there should also be a check that ensures that a problematic file made before this check is also fixed.

@maxwxyz perhaps this should also be a part of the file format version change checks?

leuriato commented:
I added the helper function and placed it at the top of `EditDatumDialog.cpp` for now.
It only checks whether the name is valid, maybe it could do more ?

Also, I added the QT_TRANSLATE_NOOP but I haven't added tne entry in `Sketcher.ts`, as I'm not sure of how it works.

As for the check to ensure an old file doesn't contains invalid contraint names, I guess it should happen when the sketch loads, but I don't know where this would be.

kadet1090 commented:
Hi, thanks for the contribution!

> Also, I added the QT_TRANSLATE_NOOP but I haven't added tne entry in `Sketcher.ts`, as I'm not sure of how it works.

`*.ts` files will be updated by maintainers later, you don't need to worry about them.

As I see that you already addressed received feedback I notified @AjinkyaDahale to take a final look. I'll approve CI workflows - after it goes green and we have confirmation from reviewer it will be good to merge.

adrianinsaval commented:
this still allows using units as names, perhaps `App::ExpressionParser::isTokenAnIndentifier(constraintName)` is a better check? or `App::ExpressionParser::isTokenAUnit(constraintName) || App::ExpressionParser::isTokenAConstant(constraintName)`

these checks should probably all be unified, see #16902 and #18567
